flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/TriangleEnumerator.java

```diff
@@ -0,0 +1,347 @@
+import org.apache.flink.types.NullValue;
+import org.apache.flink.util.Collector;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+
+/**
+ * This function returns number of triangles present in the input graph.
+ * A triangle consists of three edges that connect three vertices with each other.
+ * <p>
+ * <p>
+ * The basic algorithm works as follows:
+ * It groups all edges that share a common vertex and builds triads, i.e., triples of vertices
+ * that are connected by two edges. Finally, all triads are filtered for which no third edge exists
+ * that closes the triangle.
+ * <p>
+ * <p>
+ * For a group of <i>n</i> edges that share a common vertex, the number of built triads is quadratic <i>((n*(n-1))/2)</i>.
+ * Therefore, an optimization of the algorithm is to group edges on the vertex with the smaller output degree to
+ * reduce the number of triads.
+ * This implementation extends the basic algorithm by computing output degrees of edge vertices and
+ * grouping on edges on the vertex with the smaller degree.
+ */
+
+public class TriangleEnumerator<K extends Comparable<K>, VV, EV> implements GraphAlgorithm<K, VV, EV, DataSet<Tuple3<K,K,K>>> {
+	@Override
+	public DataSet<Tuple3<K,K,K>> run(Graph<K, VV, EV> input) throws Exception {
+
+		DataSet<Edge<K, EV>> edges = input.getEdges();
+
+		// annotate edges with degrees
+		DataSet<EdgeWithDegrees<K>> edgesWithDegrees = edges.flatMap(new EdgeDuplicator<K, EV>())
+				.groupBy(0).sortGroup(1, Order.ASCENDING).reduceGroup(new DegreeCounter<K, EV>())
+				.groupBy(EdgeWithDegrees.V1, EdgeWithDegrees.V2).reduce(new DegreeJoiner<K>());
+
+		// project edges by degrees
+		DataSet<Edge<K, NullValue>> edgesByDegree = edgesWithDegrees.map(new EdgeByDegreeProjector<K>());
+		// project edges by vertex id
+		DataSet<Edge<K, NullValue>> edgesById = edgesByDegree.map(new EdgeByIdProjector<K>());
+
+		DataSet<Tuple3<K,K,K>> triangles = edgesByDegree
+				// build triads
+				.groupBy(EdgeWithDegrees.V1).sortGroup(EdgeWithDegrees.V2, Order.ASCENDING)
+				.reduceGroup(new TriadBuilder())
+				// filter triads
+				.join(edgesById).where(Triad.V2, Triad.V3).equalTo(0, 1).with(new TriadFilter<K>());
+
+		return triangles;
+	}
+
+	/**
+	 * Emits for an edge the original edge and its switched version.
+	 */
+	private static final class EdgeDuplicator<K, EV> implements FlatMapFunction<Edge<K, EV>, Edge<K, EV>> {
+
+		@Override
+		public void flatMap(Edge<K, EV> edge, Collector<Edge<K, EV>> out) throws Exception {
+			out.collect(edge);
+			Edge<K, EV> reversed = edge.reverse();
+			out.collect(reversed);
+		}
+	}
+
+	/**
+	 * Counts the number of edges that share a common vertex.
+	 * Emits one edge for each input edge with a degree annotation for the shared vertex.
+	 * For each emitted edge, the first vertex is the vertex with the smaller id.
+	 */
+	private static final class DegreeCounter<K extends Comparable<K>, EV>
```

Comment:
As I understand, you are suggesting the signature to be
```java
private static final class EdgeDuplicator<K> implements FlatMapFunction<Edge<K, EV>, Edge<K, NullValue>>
```
But since this takes the raw edges passed by the caller `EV` is still necessary, right?